File: intersection.cpp

#include <iostream>
#include "dataset.h"
#include "util.h"

using namespace std;


void intersection() {
    cout << "Merge brands: same models with different representations!" << endl;
    for (auto brand_item: model_index) {
        unordered_set<string> del_set;
        string brand_name = brand_item.first;
        cout << "Merge brand: " << brand_name << endl;
        //<model_name, product_set>
        vector<pair<string, unordered_set<string>>> model_items =
                {brand_item.second.begin(), brand_item.second.end()};
        for (int i = 0; i < model_items.size(); i++) {
            if (del_set.find(model_items[i].first) != del_set.end()) {
                continue;
            }
            for (int j = i + 1; j < model_items.size(); j ++) {
                if (del_set.find(model_items[j].first) != del_set.end()) {
                    continue;
                }
                if(set_share_more_than(model_items[i].second, model_items[j].second)) {
                    const string& from_model_name = model_items[i].first;
                    const string& to_model_name = model_items[j].first;
                    cout << "Merge: " << model_items[j].first << " to " << model_items[i].first << endl;
                    model_index[brand_name][model_items[i].first].
                        insert(model_items[j].second.begin(), model_items[j].second.end());
                    del_set.insert(model_items[j].first);
                }
            }

        }
        for (const string& del_model: del_set) {
            model_index[brand_name].erase(del_model);
        }
    }
}

File: python version/intersection.py
import dataset

E = 2



def intersection():
    print("Merge same models with different representations")
    deleted_set = set()
    for brand in dataset.model_index.keys():
        print("Merge brand:", brand)
        model_items = list(dataset.model_index[brand].items())
        for i in range(0, len(model_items)):
            if model_items[i][0] in deleted_set:
                continue
            for j in range(i+1, len(model_items)):
                if model_items[j][0] in deleted_set:
                    continue
                if len(set(model_items[i][1]) & set(model_items[j][1])) > E:
                    # If two model have more than 2 intersection json instances, merge these two models
                    dataset.model_index[brand][model_items[i][0]] += \
                        dataset.model_index[brand][model_items[j][0]]
                    print("Merge:", model_items[j][0], model_items[i][0])
                    del dataset.model_index[brand][model_items[j][0]]
                    deleted_set.add(model_items[j][0])






